File: municipality/types/municipality.ts
/**
 * Mirrors `citizen app/backend/api/app/schemas/municipality.py` field-for-
 * field (camelCase here, snake_case there — translated automatically at
 * the network boundary, see `services/api/client.ts`/`utils/case.ts`).
 * The backend is the source of truth; this is a hand-maintained mirror.
 */
import type { HazardStatusType, HazardTypeType, InspectionDecisionType, RepairPriorityType, RepairStatusType, SeverityType, VerificationMethodType, VerificationStateType } from '@/constants/enums';

export interface City {
  id: string;
  code: string;
  name: string;
  state: string | null;
  centerLatitude: number | null;
  centerLongitude: number | null;
  wardCount: number;
}

export interface Ward {
  id: string;
  cityId: string;
  code: string;
  name: string;
}

export interface MunicipalityOfficer {
  id: string;
  fullName: string;
  email: string;
  role: string;
  officerRole: string;
  municipalityId: string;
  municipalityCode: string;
  municipalityName: string;
  allowedCityIds: string[];
  defaultCityId: string | null;
  wardIds: string[] | null;
  permissions: string[];
}

export interface MunicipalityMeResponse {
  officer: MunicipalityOfficer;
  authorizedCities: City[];
}

export interface Priority {
  label: SeverityType;
  score: number;
  reasons: string[];
  recommendation: string;
}

export interface MunicipalityHazard {
  id: string;
  hazardCode: string;
  type: HazardTypeType;
  latitude: number;
  longitude: number;
  locationText: string;
  roadName: string | null;
  cityId: string | null;
  wardId: string | null;
  wardName: string | null;
  severity: SeverityType;
  status: HazardStatusType;
  aiConfidence: number | null;
  imageUrl: string | null;
  description: string | null;
  sourceSummary: string;
  repairStatus: RepairStatusType | null;
  repairCode: string | null;
  createdAt: string;
  updatedAt: string;
  lastObservedAt: string;
}

export interface MunicipalityHazardListResponse {
  items: MunicipalityHazard[];
  total: number;
  page: number;
  pageSize: number;
}

export interface CitizenReportEvidence {
  id: string;
  reportCode: string;
  hazardType: HazardTypeType;
  severity: SeverityType;
  description: string | null;
  media: string[];
  createdAt: string;
}

export interface AIAnalysisEvidence {
  id: string;
  detected: boolean;
  confidence: number;
  severity: SeverityType | null;
  modelVersion: string;
  imageUrl: string;
  createdAt: string;
}

export interface FleetObservation {
  id: string;
  vehicleId: string;
  observedAt: string;
  observationState: string | null;
  confidence: number | null;
  severity: SeverityType | null;
  imageUrl: string | null;
  latitude: number;
  longitude: number;
  dataQuality: string | null;
}

export interface MunicipalityAction {
  id: string;
  actionType: string;
  notes: string | null;
  performedBy: string | null;
  fromStatus: string | null;
  toStatus: string | null;
  createdAt: string;
}

export interface CurrentRepairSummary {
  id: string;
  repairCode: string;
  status: RepairStatusType;
  priority: RepairPriorityType;
  team: string | null;
  targetDate: string | null;
}

export interface MunicipalityHazardDetail extends MunicipalityHazard {
  citizenReports: CitizenReportEvidence[];
  latestAiAnalysis: AIAnalysisEvidence | null;
  latestFleetObservations: FleetObservation[];
  municipalityActions: MunicipalityAction[];
  currentRepair: CurrentRepairSummary | null;
  priority: Priority;
  resolutionState: 'ACTIVE' | 'RESOLVED' | 'REOPENED';
  reopenedFromHazardId: string | null;
}

export interface Verification {
  hazardId: string;
  state: VerificationStateType;
  confidence: number;
  detectedCount: number;
  clearCount: number;
  observations: FleetObservation[];
  summary: string;
  previouslyResolved: boolean;
  reopenSuggested: boolean;
}

export interface TimelineEvent {
  id: string;
  label: string;
  detail: string | null;
  fromStatus: string | null;
  toStatus: string | null;
  actor: string | null;
  createdAt: string;
}

export interface TimelineResponse {
  hazardId: string;
  events: TimelineEvent[];
}

export interface ResolveHazardPayload {
  resolutionNotes: string;
  repairId?: string | null;
  verificationMethod?: VerificationMethodType;
  latitude?: number | null;
  longitude?: number | null;
  evidenceMediaIds?: string[];
}

export interface Resolution {
  id: string;
  hazardId: string;
  repairId: string | null;
  verifiedBy: string;
  verifiedAt: string;
  resolutionNotes: string | null;
  verificationMethod: VerificationMethodType;
  status: string;
}

export interface RepairCreatePayload {
  hazardId: string;
  department: string;
  zone?: string | null;
  team?: string | null;
  assignedOfficerName?: string | null;
  priority?: RepairPriorityType;
  targetDate?: string | null;
  notes?: string | null;
}

export interface Repair {
  id: string;
  repairCode: string;
  hazardId: string;
  hazardRoadName: string | null;
  hazardLocationText: string | null;
  cityId: string;
  department: string;
  zone: string | null;
  team: string | null;
  assignedOfficerName: string | null;
  priority: RepairPriorityType;
  targetDate: string | null;
  status: RepairStatusType;
  notes: string | null;
  startedAt: string | null;
  completedAt: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface RepairListResponse {
  items: Repair[];
  total: number;
}

export interface RepairProgressPayload {
  note?: string | null;
  photoUrl?: string | null;
  latitude?: number | null;
  longitude?: number | null;
}

export interface RepairProgress {
  id: string;
  note: string | null;
  photoUrl: string | null;
  latitude: number | null;
  longitude: number | null;
  createdByName: string | null;
  createdAt: string;
}

export interface RepairDetail extends Repair {
  progressEntries: RepairProgress[];
  timeline: TimelineEvent[];
}

export interface InspectionDecisionPayload {
  decision: InspectionDecisionType;
  notes?: string | null;
}

export interface Inspection {
  id: string;
  repairId: string;
  hazardId: string;
  inspectorId: string;
  decision: InspectionDecisionType;
  notes: string | null;
  createdAt: string;
}

export interface PriorityRoad {
  roadName: string;
  priority: SeverityType;
  activeHazardCount: number;
  criticalHazardCount: number;
  reason: string;
}

export interface RecentActivityItem {
  id: string;
  label: string;
  detail: string | null;
  hazardId: string | null;
  createdAt: string;
}

export interface FleetCoverageSummary {
  observedTodayKmEstimate: number;
  coveragePercent: number;
  dataGapSegments: number;
}

export interface Dashboard {
  cityId: string;
  cityName: string;
  activeHazards: number;
  criticalHazards: number;
  underRepair: number;
  resolvedToday: number;
  priorityRoads: PriorityRoad[];
  recentActivity: RecentActivityItem[];
  fleetCoverage: FleetCoverageSummary;
  generatedAt: string;
}

export interface AnalyticsSummary {
  activeHazards: number;
  criticalHazards: number;
  resolvedThisWeek: number;
  avgResolutionTimeHours: number | null;
  recurringHazardRoads: number;
  citizenReportsCount: number;
  fleetObservationCount: number;
  avgRepairTurnaroundHours: number | null;
}

export interface WardAnalyticsItem {
  wardId: string | null;
  wardName: string;
  activeHazards: number;
  criticalHazards: number;
  resolvedHazards: number;
}

export interface SeverityAnalyticsItem {
  severity: SeverityType;
  count: number;
}

export interface ResolutionTrendPoint {
  date: string;
  resolvedCount: number;
}

export interface RecurringHazardItem {
  roadName: string;
  reportCount: number;
  recurringEventCount: number;
  recommendation: string;
}

export interface FleetCoverage {
  observedTodayKmEstimate: number;
  coveragePercent: number;
  dataGapSegments: number;
  roadsObservedToday: number;
  totalRoadsTracked: number;
  staleRoads: string[];
}

File: citizen_app/backend/api/app/schemas/municipality.py
"""
Pydantic response/request shapes for the Municipality app
(`app/api/v1/municipality/*`). Kept in one file, unlike the citizen
schemas' one-file-per-resource split, because the Municipality API surface
is one bounded module added on top of an existing app (see
`docs/architecture.md`) — this mirrors how `app/api/v1/municipality/` is
laid out as a subpackage rather than flat files.
"""
import uuid
from datetime import date, datetime

from pydantic import BaseModel, EmailStr, Field

from app.models.enums import (
    HazardStatus,
    HazardType,
    InspectionDecision,
    RepairPriority,
    RepairStatus,
    Severity,
    VerificationMethod,
    VerificationState,
)

# --------------------------------------------------------------------------
# Auth / profile / city context (sections 08/13/14)
# --------------------------------------------------------------------------


class MunicipalityLoginRequest(BaseModel):
    municipality_code: str = Field(min_length=2, max_length=32, description='e.g. "MUN-CHN"')
    email: EmailStr
    password: str = Field(min_length=8, max_length=128)


class TokenPair(BaseModel):
    access_token: str
    refresh_token: str
    token_type: str = "bearer"


class CityOut(BaseModel):
    id: uuid.UUID
    code: str
    name: str
    state: str | None
    center_latitude: float | None
    center_longitude: float | None
    ward_count: int = 0

    model_config = {"from_attributes": True}


class WardOut(BaseModel):
    id: uuid.UUID
    city_id: uuid.UUID
    code: str
    name: str

    model_config = {"from_attributes": True}


class MunicipalityOfficerOut(BaseModel):
    id: uuid.UUID
    full_name: str
    email: EmailStr
    role: str = "MUNICIPALITY_OFFICER"
    officer_role: str
    municipality_id: uuid.UUID
    municipality_code: str
    municipality_name: str
    allowed_city_ids: list[uuid.UUID]
    default_city_id: uuid.UUID | None
    ward_ids: list[str] | None
    permissions: list[str]


class MunicipalityAuthResponse(BaseModel):
    officer: MunicipalityOfficerOut
    tokens: TokenPair


class MunicipalityMeResponse(BaseModel):
    officer: MunicipalityOfficerOut
    authorized_cities: list[CityOut]


# --------------------------------------------------------------------------
# Hazards (sections 17-32, 74)
# --------------------------------------------------------------------------


class PriorityOut(BaseModel):
    """Decision-support output only — never the raw formula (section 31/76)."""

    label: Severity
    score: int = Field(ge=0, le=100)
    reasons: list[str]
    recommendation: str


class MunicipalityHazardOut(BaseModel):
    id: uuid.UUID
    hazard_code: str
    type: HazardType
    latitude: float
    longitude: float
    location_text: str
    road_name: str | None
    city_id: uuid.UUID | None
    ward_id: uuid.UUID | None
    ward_name: str | None = None
    severity: Severity
    status: HazardStatus
    ai_confidence: float | None
    image_url: str | None
    description: str | None
    source_summary: str
    repair_status: RepairStatus | None = None
    repair_code: str | None = None
    created_at: datetime
    updated_at: datetime
    last_observed_at: datetime

    model_config = {"from_attributes": True}


class MunicipalityHazardListResponse(BaseModel):
    items: list[MunicipalityHazardOut]
    total: int
    page: int
    page_size: int


class CitizenReportEvidenceOut(BaseModel):
    id: uuid.UUID
    report_code: str
    hazard_type: HazardType
    severity: Severity
    description: str | None
    media: list[str]
    created_at: datetime


class AIAnalysisEvidenceOut(BaseModel):
    id: uuid.UUID
    detected: bool
    confidence: float
    severity: Severity | None
    model_version: str
    image_url: str
    created_at: datetime


class FleetObservationOut(BaseModel):
    id: uuid.UUID
    vehicle_id: str
    observed_at: datetime
    observation_state: str | None
    confidence: float | None
    severity: Severity | None
    image_url: str | None
    latitude: float
    longitude: float
    data_quality: str | None


class MunicipalityActionOut(BaseModel):
    id: uuid.UUID
    action_type: str
    notes: str | None
    performed_by: str | None
    from_status: str | None
    to_status: str | None
    created_at: datetime


class CurrentRepairSummaryOut(BaseModel):
    id: uuid.UUID
    repair_code: str
    status: RepairStatus
    priority: RepairPriority
    team: str | None
    target_date: date | None


class MunicipalityHazardDetailOut(MunicipalityHazardOut):
    citizen_reports: list[CitizenReportEvidenceOut]
    latest_ai_analysis: AIAnalysisEvidenceOut | None
    latest_fleet_observations: list[FleetObservationOut]
    municipality_actions: list[MunicipalityActionOut]
    current_repair: CurrentRepairSummaryOut | None
    priority: PriorityOut
    resolution_state: str  # "ACTIVE" | "RESOLVED" | "REOPENED"
    reopened_from_hazard_id: uuid.UUID | None = None


class VerificationOut(BaseModel):
    hazard_id: uuid.UUID
    state: VerificationState
    confidence: int = Field(ge=0, le=100)
    detected_count: int
    clear_count: int
    observations: list[FleetObservationOut]
    summary: str
    previously_resolved: bool = False
    reopen_suggested: bool = False


class TimelineEventOut(BaseModel):
    id: uuid.UUID
    label: str
    detail: str | None
    from_status: str | None
    to_status: str | None
    actor: str | None
    created_at: datetime


class TimelineResponse(BaseModel):
    hazard_id: uuid.UUID
    events: list[TimelineEventOut]


class ResolveHazardRequest(BaseModel):
    resolution_notes: str = Field(min_length=1, max_length=2000)
    repair_id: uuid.UUID | None = None
    verification_method: VerificationMethod = VerificationMethod.MUNICIPAL_INSPECTION
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)
    evidence_media_ids: list[str] = Field(default_factory=list)


class ResolutionOut(BaseModel):
    id: uuid.UUID
    hazard_id: uuid.UUID
    repair_id: uuid.UUID | None
    verified_by: uuid.UUID
    verified_at: datetime
    resolution_notes: str | None
    verification_method: VerificationMethod
    status: str

    model_config = {"from_attributes": True}


class ReopenHazardRequest(BaseModel):
    note: str | None = Field(default=None, max_length=500)


# --------------------------------------------------------------------------
# Repairs / progress / inspections (sections 33-40)
# --------------------------------------------------------------------------


class RepairCreateRequest(BaseModel):
    hazard_id: uuid.UUID
    department: str = Field(min_length=1, max_length=120)
    zone: str | None = Field(default=None, max_length=120)
    team: str | None = Field(default=None, max_length=120)
    assigned_officer_name: str | None = Field(default=None, max_length=120)
    priority: RepairPriority = RepairPriority.MEDIUM
    target_date: date | None = None
    notes: str | None = Field(default=None, max_length=2000)


class RepairOut(BaseModel):
    id: uuid.UUID
    repair_code: str
    hazard_id: uuid.UUID
    hazard_road_name: str | None = None
    hazard_location_text: str | None = None
    city_id: uuid.UUID
    department: str
    zone: str | None
    team: str | None
    assigned_officer_name: str | None
    priority: RepairPriority
    target_date: date | None
    status: RepairStatus
    notes: str | None
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class RepairListResponse(BaseModel):
    items: list[RepairOut]
    total: int


class RepairProgressCreateRequest(BaseModel):
    note: str | None = Field(default=None, max_length=2000)
    photo_url: str | None = None
    latitude: float | None = Field(default=None, ge=-90, le=90)
    longitude: float | None = Field(default=None, ge=-180, le=180)


class RepairProgressOut(BaseModel):
    id: uuid.UUID
    note: str | None
    photo_url: str | None
    latitude: float | None
    longitude: float | None
    created_by_name: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


class RepairDetailOut(RepairOut):
    progress_entries: list[RepairProgressOut]
    timeline: list[TimelineEventOut]


class InspectionDecisionRequest(BaseModel):
    decision: InspectionDecision
    notes: str | None = Field(default=None, max_length=2000)


class InspectionOut(BaseModel):
    id: uuid.UUID
    repair_id: uuid.UUID
    hazard_id: uuid.UUID
    inspector_id: uuid.UUID
    decision: InspectionDecision
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}


# --------------------------------------------------------------------------
# Dashboard (sections 15/16)
# --------------------------------------------------------------------------


class PriorityRoadOut(BaseModel):
    road_name: str
    priority: Severity
    active_hazard_count: int
    critical_hazard_count: int
    reason: str


class RecentActivityItemOut(BaseModel):
    id: uuid.UUID
    label: str
    detail: str | None
    hazard_id: uuid.UUID | None
    created_at: datetime


class FleetCoverageSummaryOut(BaseModel):
    observed_today_km_estimate: float
    coverage_percent: int
    data_gap_segments: int


class DashboardOut(BaseModel):
    city_id: uuid.UUID
    city_name: str
    active_hazards: int
    critical_hazards: int
    under_repair: int
    resolved_today: int
    priority_roads: list[PriorityRoadOut]
    recent_activity: list[RecentActivityItemOut]
    fleet_coverage: FleetCoverageSummaryOut
    generated_at: datetime


# --------------------------------------------------------------------------
# Analytics (sections 44-47, 68)
# --------------------------------------------------------------------------


class AnalyticsSummaryOut(BaseModel):
    active_hazards: int
    critical_hazards: int
    resolved_this_week: int
    avg_resolution_time_hours: float | None
    recurring_hazard_roads: int
    citizen_reports_count: int
    fleet_observation_count: int
    avg_repair_turnaround_hours: float | None


class WardAnalyticsItemOut(BaseModel):
    ward_id: uuid.UUID | None
    ward_name: str
    active_hazards: int
    critical_hazards: int
    resolved_hazards: int


class SeverityAnalyticsItemOut(BaseModel):
    severity: Severity
    count: int


class ResolutionTrendPointOut(BaseModel):
    date: date
    resolved_count: int


class RecurringHazardItemOut(BaseModel):
    road_name: str
    report_count: int
    recurring_event_count: int
    recommendation: str


class FleetCoverageOut(BaseModel):
    observed_today_km_estimate: float
    coverage_percent: int
    data_gap_segments: int
    roads_observed_today: int
    total_roads_tracked: int
    stale_roads: list[str]


# --------------------------------------------------------------------------
# Notifications — municipality reuses `GET /api/notifications` (same table,
# scoped by JWT `user_id`, already secure). No municipality-specific
# notification schema is needed; see docs/architecture.md.
# --------------------------------------------------------------------------
